AX: Add initial ariaNotify implementation

hoffmanjoshua opened the pull request:
#### 997691b638cb8da31793e0f9f70ee43b8ea6bf49
<pre>
AX: Add initial ariaNotify implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=302136">https://bugs.webkit.org/show_bug.cgi?id=302136</a>
<a href="https://rdar.apple.com/164216162">rdar://164216162</a>

Reviewed by Tyler Wilcock.

This adds initial support for the forthcoming ariaNotify spec on mac and iOS platforms.

On both platforms, this adds a new javascript interface to both Document and Element:
- ariaNotify(&quot;Message&quot;)
- ariaNotify(&quot;Message&quot;, { priority: &quot;high&quot;, interrupt: &quot;pending&quot; })

The intent of these new functions is for web authors to send announcements to screen readers
with specified priority and interruption behavior.

To do this, when this method is called, we now send announcement requested notifications
on both platforms that include both the priority and interrupt behavior specified, alongside
the announcement&apos;s text and language.

On Mac, we directly send this from WebContent. On iOS, we forward this information to the
UI process which handles constructing and sending the notification to the system. This is
done via IPC and a new struct, which packages up the bits that we need to accurately
build the attributed string that the AX notification system will handle.

For testing purposes, we re-post this notification in a way that the tests can parse.

Three new tests (2 mac, 1 iOS) were added to exercise this behavior, and validate both the
document and element implementations.

Tests: accessibility/ios-simulator/aria-notify.html
       accessibility/mac/aria-notify-with-options.html
       accessibility/mac/aria-notify.html

* LayoutTests/accessibility/ios-simulator/aria-notify-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/aria-notify.html: Added.
* LayoutTests/accessibility/mac/aria-notify-expected.txt: Added.
* LayoutTests/accessibility/mac/aria-notify-with-options.html: Added.
* LayoutTests/accessibility/mac/aria-notify.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/accessibility/mac/aria-notify-with-options-expected.txt: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::setShouldRepostNotificationsForTests):
(WebCore::AXObjectCache::postARIANotifyNotification):
(WebCore::AXObjectCache::handleAttributeChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/cocoa/AXObjectCacheCocoa.mm: Added.
(WebCore::notifyPriorityToAXValueString):
(WebCore::interruptBehaviorToAXValueString):
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::postPlatformARIANotifyNotification):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::visibleCharacterRange const):
(WebCore::AXIsolatedObject::textMarkerRangeForNSRange const):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXPostNotificationWithUserInfo):
(WebCore::AXObjectCache::postPlatformAnnouncementNotification):
(WebCore::AXObjectCache::postPlatformARIANotifyNotification):
(WebCore::AXObjectCache::platformHandleFocusedUIElementChanged):
(WebCore::AXObjectCache::setShouldRepostNotificationsForTests): Deleted.
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(+[WebAccessibilityObjectWrapperBase accessibilitySetShouldRepostNotifications:]):
(isValueTypeSupported):
* Source/WebCore/dom/AriaNotifyOptions.h: Added.
* Source/WebCore/dom/AriaNotifyOptions.idl: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ariaNotify):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::ariaNotify):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::relayAriaNotifyNotification const):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::relayAriaNotifyNotification):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::relayAriaNotifyNotification):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebChromeClientIOS.mm:
(WebKit::WebChromeClient::relayAriaNotifyNotification const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::relayAriaNotifyNotification):
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:
* Tools/WebKitTestRunner/InjectedBundle/cocoa/AccessibilityCommonCocoa.mm:
(WTR::makeValueRefForValue):

Canonical link: <a href="https://commits.webkit.org/302795@main">https://commits.webkit.org/302795@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a2952c8b863dd38a0261e4969004702264a1083

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130054 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2315 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40915 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137449 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81577 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d9ff9bb8-69fb-44a1-bdb9-a0506f934b4f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131925 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2282 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2207 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99069 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66872 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f57d42c5-85b7-43dc-9ba2-808879f1b2b1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1710 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116487 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79770 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1382e6d3-c500-4cab-8a95-0517205c202d) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/129404 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1750 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34614 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80721 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122046 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/110144 "Build was cancelled. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests (cancelled)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35122 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139928 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/128506 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2109 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1954 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107575 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2154 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112834 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107468 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27394 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1683 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31284 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54944 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2182 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65551 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/161520 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1999 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40263 "Passed tests") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/2205 "Build was cancelled. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (cancelled)") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2105 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/ios-simulator/aria-notify.html:
```diff
@@ -0,0 +1,36 @@
+if (window.accessibilityController) {
+    window.jsTestIsAsync = true;
+
+    accessibilityController.addNotificationListener(notificationCallback);
```
I think we need to remove the notification listener before finishing the test

twilco commented on LayoutTests/accessibility/mac/aria-notify-with-options.html:
```diff
@@ -0,0 +1,52 @@
+if (accessibilityController) {
+    window.jsTestIsAsync = true;
+
+    accessibilityController.addNotificationListener(notificationCallback);
```
I think we need to remove the notification listener before finishing the test

twilco commented on LayoutTests/accessibility/mac/aria-notify-with-options.html:
```diff
@@ -0,0 +1,52 @@
+
+    if (notificationCount == 3) {
+        document.getElementById("button").ariaNotify("Boston", { priority: "high" });
+    }
```
These don't need curly braces I think

twilco commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -129,6 +135,18 @@ struct AXDebugInfo {
+struct AriaNotifyData {
+    String message;
+    NotifyPriority priority;
+    InterruptBehavior interrupt;
```
I think it's important to define a default value for enum values, as otherwise it can cause a security vulnerability if these fields are referenced without being initialized first.

e.g. `NotifyPriority priority { ... };`

hoffmanjoshua replied:
Good call, added these in

twilco commented on Source/WebCore/accessibility/cocoa/AXObjectCacheCocoa.mm:
```diff
@@ -0,0 +1,54 @@
+
+RetainPtr<NSString> notifyPriorityToString(NotifyPriority priority)
+{
+    return priority == NotifyPriority::Normal ? @"Normal" : @"High";
```
Should we use a switch so the code stops compiling if we add a new enum value?

hoffmanjoshua replied:
switched! (pun intended)

minorninth commented on LayoutTests/platform/mac/accessibility/mac/aria-notify-with-options-expected.txt:
```diff
@@ -0,0 +1,31 @@
+AnnouncementKey: Boston
+AXARIAAnnouncementInterruptBehavior: None
+AXARIAAnnouncementPriority: High
+AXAnnouncementLangaugeKey: da
```
Language is misspelled here; please check for the misspelling throughout the patch

hoffmanjoshua replied:
Thanks for catching—updated

minorninth commented on Source/WebCore/accessibility/cocoa/AXObjectCacheCocoa.mm:
```diff
@@ -0,0 +1,54 @@
+
+namespace WebCore {
+
+RetainPtr<NSString> notifyPriorityToString(NotifyPriority priority)
```
Many fooToString methods are used for logging / debugging. I wonder if these should be a little more explicit about the fact that these are the the string values of userInfo keys for the official AX API?

Something like notifyPriorityToAXNotificationValueString? Or maybe you can come up with better.

hoffmanjoshua replied:
Good idea! Renamed to `notifyPriorityToAXValueString` to be a little more considerate but still communicate the same idea that this is not for testing but an actual value.

minorninth commented on Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
```diff
@@ -133,6 +133,26 @@
+        page->chrome().relayAriaNotifyNotification(WTFMove(notificationData));
+
+    // For LayoutTests on the simulator, also call the wrapper's accessibilityPostedNotification.
+    if (RefPtr root = getOrCreate(m_document->view())) {
```
This is a fair amount of work to do only for tests. How about checking accessibilitySetShouldRepostNotifications first (it's static) and doing an early exit if it's false?

hoffmanjoshua replied:
Moved some stuff around (see new global var) but fixed this

minorninth reviewed (approved):
lgtm

hoffmanjoshua committed:
AX: Add initial ariaNotify implementation
https://bugs.webkit.org/show_bug.cgi?id=302136
rdar://164216162

Reviewed by Tyler Wilcock.

This adds initial support for the forthcoming ariaNotify spec on mac and iOS platforms.

On both platforms, this adds a new javascript interface to both Document and Element:
- ariaNotify("Message")
- ariaNotify("Message", { priority: "high", interrupt: "pending" })

The intent of these new functions is for web authors to send announcements to screen readers
with specified priority and interruption behavior.

To do this, when this method is called, we now send announcement requested notifications
on both platforms that include both the priority and interrupt behavior specified, alongside
the announcement's text and language.

On Mac, we directly send this from WebContent. On iOS, we forward this information to the
UI process which handles constructing and sending the notification to the system. This is
done via IPC and a new struct, which packages up the bits that we need to accurately
build the attributed string that the AX notification system will handle.

For testing purposes, we re-post this notification in a way that the tests can parse.

Three new tests (2 mac, 1 iOS) were added to exercise this behavior, and validate both the
document and element implementations.

Tests: accessibility/ios-simulator/aria-notify.html
       accessibility/mac/aria-notify-with-options.html
       accessibility/mac/aria-notify.html

* LayoutTests/accessibility/ios-simulator/aria-notify-expected.txt: Added.
* LayoutTests/accessibility/ios-simulator/aria-notify.html: Added.
* LayoutTests/accessibility/mac/aria-notify-expected.txt: Added.
* LayoutTests/accessibility/mac/aria-notify-with-options.html: Added.
* LayoutTests/accessibility/mac/aria-notify.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/accessibility/mac/aria-notify-with-options-expected.txt: Added.
* LayoutTests/resources/accessibility-helper.js:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::setShouldRepostNotificationsForTests):
(WebCore::AXObjectCache::postARIANotifyNotification):
(WebCore::AXObjectCache::handleAttributeChange):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/cocoa/AXObjectCacheCocoa.mm: Added.
(WebCore::notifyPriorityToAXValueString):
(WebCore::interruptBehaviorToAXValueString):
* Source/WebCore/accessibility/ios/AXObjectCacheIOS.mm:
(WebCore::AXObjectCache::postPlatformARIANotifyNotification):
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::visibleCharacterRange const):
(WebCore::AXIsolatedObject::textMarkerRangeForNSRange const):
* Source/WebCore/accessibility/mac/AXObjectCacheMac.mm:
(WebCore::AXPostNotificationWithUserInfo):
(WebCore::AXObjectCache::postPlatformAnnouncementNotification):
(WebCore::AXObjectCache::postPlatformARIANotifyNotification):
(WebCore::AXObjectCache::platformHandleFocusedUIElementChanged):
(WebCore::AXObjectCache::setShouldRepostNotificationsForTests): Deleted.
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
(+[WebAccessibilityObjectWrapperBase accessibilitySetShouldRepostNotifications:]):
(isValueTypeSupported):
* Source/WebCore/dom/AriaNotifyOptions.h: Added.
* Source/WebCore/dom/AriaNotifyOptions.idl: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ariaNotify):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Document.idl:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::ariaNotify):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/Element.idl:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::relayAriaNotifyNotification const):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::relayAriaNotifyNotification):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::relayAriaNotifyNotification):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/ios/WebChromeClientIOS.mm:
(WebKit::WebChromeClient::relayAriaNotifyNotification const):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::relayAriaNotifyNotification):
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:
* Tools/WebKitTestRunner/InjectedBundle/cocoa/AccessibilityCommonCocoa.mm:
(WTR::makeValueRefForValue):

Canonical link: https://commits.webkit.org/302795@main
